### ISSUE TITLE
Profile TW performance with large task DBs

### ISSUE THREAD

**Issue by djmitche** [GothenburgBitFactory/taskwarrior#3618]
...Probably running a profiler is the action item from this issue...

_Originally posted by @djmitche in https://github.com/GothenburgBitFactory/taskwarrior/issues/3329#issuecomment-2335007219_

---

I think this would be best done from Taskwarrior, treating the Rust code as a "black box". It would be great to learn which calls into Rust are taking the longest -- and what non-Rust things are taking a long time, too. I can think of a few possibilities we might learn, but the truth is probably something different:
 * Opening the DB might take a long time (partly due to #3418)
 * Writes may be super-slow (https://github.com/GothenburgBitFactory/taskchampion/issues/373 should help with this, by doing writes in fewer transactions)
 * We may be calling into Rust repeatedly in a way that is wasteful (e.g., `TDB2::pending_tasks`)
 * The translation between Rust and C++ data structures may be slow (C++'s `std::string` and Rust's `String` are not compatible)

I don't know much about profiling in C++, but perhaps you, dear reader, do?

**Comment by felixschurk** [comment]
I think in order to have it somehow comparable we should use here the existing `load, run_perf` tasks which are in the `performance` folder. But probably (at least for a first run) in a bit cut down size.. Currently it would create 8510 tasks, which is fairly large :D

I am planning to take a look into this, however I will probably not have time before end of September :/

**Comment by felixschurk** [comment]
I did let it run on my local machine which runs on Fedora, having an i5-1145G7 processor and 16 GiB of RAM. Not using the machine for anything else super intense at the profiling time and then it came up with some call graph like this:
![image](https://github.com/user-attachments/assets/0d6e75ef-001d-4ead-9ea8-8a35490acdf0)

At least based on that what takes the most of the time is the rendering, starting, and the rebuilding also the working set of taskwarrior.

I am not really sure how the best way to share results of performance profiling etc. 

I also did let run the same performance suite (the one which is with taskwarrior) on the release of 2.6.2 and there the total runtime was only 2.7s vs. 4.25s on the current develop branch. I mean the measurements are not about the current absolute time nor anything (I only did let them run once so far, so in order to have there something super consistent multiple runs and then some average or so would be needed) but still one sees that taskwarrior 3 is noticeably slower, but solely from the profiling I can not really pin it down to one method where the problem might be coming from.

The performance script also returns some counters for garbage collection etc. and there the counters are an order more for the taskwarrior 3, but I am not aware on how they are computed so might also report something wrong.

**Comment by felixschurk** [comment]
Okay, I just noticed that one "counter" or timer was lost due to the transition to the TDB2 backend. More specifically, the timer which tracked how long it takes to load some task set into Taskwarrior. 

The counter was in Line 419 of TDB2.cpp.
https://github.com/GothenburgBitFactory/taskwarrior/commit/5bb98579841d41e680280ada7bf3a3e9a9940553#diff-d6d65d1f59ba33ef5268d5bc8a49ed26f397b899b902332404aee34e800cbc0bL419

```cpp
Context::getContext ().time_load_us += timer.total_us ();
```

This explains why the reporting of the performance always states `load:0` for the latest taskwarrior.

**Comment by djmitche** [comment]
That flame graph seems to have multiple commands in it -- is it a concatenation from multiple runs? It's a little difficult to interpret the graphic, since things are cut off. Is there a way to share the source file from which that was drawn?

https://github.com/GothenburgBitFactory/taskwarrior/pull/3635 should restore the `load` timer.

**Comment by felixschurk** [comment]
I will come back and see how to extract it better to a simple output.

**Comment by felixschurk** [comment]
Okay, I finally found some time to look a bit back into it. 

I run the `next` report on the performance database once and got an overview of the call graph with [callgrind](https://valgrind.org/docs/manual/cl-manual.html)
```
Category                   Data          
Pending                    2128          
Waiting                    0
Recurring                  0             
Completed                  5957
Deleted                    0             
Total                      8085
Annotations                425           
Unique tags                292
Projects                   230           
Blocked tasks              0
Blocking tasks             0             
Undo transactions          8510
Sync backlog transactions  71007         
Tasks tagged               99.2%
Oldest task                2024-11-15    
Newest task                2024-11-15
Task used for              4min          
Task added every
Task completed every                     
Average time pending       21min
Average desc length        125 characters
```

The results can be seen in the following pdf, however it is a bit messy and mainly shows there the counter of the calls. However, reproducing it following the manual of callgrind is quite straight forward.
For viewing of the attached file I would recommend [kcachegrind](https://kcachegrind.github.io/html/Home.html).

Long story short, I know to little of the code in TW to properly understand the output or really get an overview which call might be unnecessary or what really limits the performance. I will try to understand and wrap my head around it.

On my laptop, using the same databse it takes approx 0,2 s with `v2.6.2` and 0,6s with `v3.2`, meaning we are currently still three times slower than `v2.6.2`.

[call-graph.pdf](https://github.com/user-attachments/files/17780883/call-graph.pdf)
[callgrind.out.zip](https://github.com/user-attachments/files/17780967/callgrind.out.zip)

**Comment by djmitche** [comment]
This suggests that most of that time is spent in the `gc` operation, and the included PDF agrees: gc is basically rebuilding the working set, and a lot of the calls are doing that. So that's probably a good place to start! https://github.com/GothenburgBitFactory/taskchampion/issues/414 may be a start there, but there is probably room for more.
![image](https://github.com/user-attachments/assets/44cddfee-f809-4805-a31d-2a936b03e77d)

Here's the code for rebuilding the working set:
```rust
/// Rebuild the working set using a function to identify tasks that should be in the set.  This
/// renumbers the existing working-set tasks to eliminate gaps, and also adds any tasks that
/// are not already in the working set but should be.  The rebuild occurs in a single
/// trasnsaction against the storage backend.
pub fn rebuild<F>(txn: &mut dyn StorageTxn, in_working_set: F, renumber: bool) -> Result<()>
where                                   
    F: Fn(&TaskMap) -> bool,
{                                                       
    let mut new_ws = vec![None]; // index 0 is always None
    let mut seen = HashSet::new();
                                                                               
    // The goal here is for existing working-set items to be "compressed' down to index 1, so
    // we begin by scanning the current working set and inserting any tasks that should still
    // be in the set into new_ws, implicitly dropping any tasks that are no longer in the
    // working set.             
    for elt in txn.get_working_set()?.drain(1..) {
        if let Some(uuid) = elt {
            if let Some(task) = txn.get_task(uuid)? {
                if in_working_set(&task) {    
                    new_ws.push(Some(uuid));
                    seen.insert(uuid);               
                    continue;
                }                           
            }                        
        }
                                                                               
        // if we are not renumbering, then insert a blank working-set entry here
        if !renumber {
            new_ws.push(None);
        }                 
    }    

    // if renumbering, clear the working set and re-add
    if renumber {             
        txn.clear_working_set()?;                                     
        for elt in new_ws.drain(1..new_ws.len()).flatten() {
            txn.add_to_working_set(elt)?;
        }       
    } else {                           
        // ..otherwise, just clear the None items determined above from the working set
        for (i, elt) in new_ws.iter().enumerate().skip(1) {
            if elt.is_none() {         
                txn.set_working_set_item(i, None)?;
            }            
        }        
    }         
                                                                               
    // Now go hunting for tasks that should be in this list but are not, adding them at the
    // end of the list, whether renumbering or not
    for (uuid, task) in txn.all_tasks()? {
        if !seen.contains(&uuid) && in_working_set(&task) {                                                                                                    
            txn.add_to_working_set(uuid)?;            
        }                                                             
    }           
                                                                                                                                                               
    txn.commit()?;                        
    Ok(())                                                                  
}         
```

Drilling down into the working set rebuild, the majority of the time is in `all_tasks` (called once at the end but returning a lot of data) and `get_task` (called once for each task in the existing working set). There's a little time spent in adding tasks to the working set (perhaps fixed by the issue linked above), but for the most part it's the fetch-all-tasks bit that is taking a lot of time.
![image](https://github.com/user-attachments/assets/ebbb7874-9780-4a0b-bdca-901ce12c8f75)

Going back to `main` and drilling down through the initialization, I see that `CmdCustom::execute` (the "custom report" command) calls `handleUntil` which handles deleting pending tasks with `until` dates in the past. It takes a while because it calls `pending_tasks` to fetch all pending tasks from the DB. This is probably faster due to #3632. But, this is surprising -- that's a kind of garbage-collection, but is not guarded by `rc:gc`! This function calls `pending_tasks`, but we cache the pending tasks, which is probably called by the custom report later, so that's not wasted time.
https://github.com/GothenburgBitFactory/taskwarrior/blob/01ced3238e87b71647fcdf8179907524e97ff54a/src/TDB2.cpp#L267-L286

A few thoughts off the top of my head:
 - Maybe we could cache some kind of "no changes since last working-set rebuild" flag that would allow us to make it a no-op
    - This might be useful mixed with the ability to open a replica in read-only mode: TW could open in read-only mode by default, see if a working-set rebuild is required, and only re-open in read-write mode if it is required. 
 - We could probably do a better job of this at the DB level. The only data we need from the DB is the set of uuids for tasks with `status:pending` - that's enough to know what to add and remove from the working-set table.
 - The `handleUntil` check should be part of `gc`. It will almost always make no changes, so the same read-only/read-write considerations apply.

What do you think?

**Comment by ChargingBulle** [comment]
I feel like this should have been solved before release V3. 

Every operation takes approx 1 second per task and is **significantly** slower than the V2:
<details>

task 3.4.2 built for Linux

```
cbulle@debian-trixie:~$ time task 3 mod wait:mon
Modifying task 3 'hidden'.
Modified 1 task.
TASKDATA override: /tmp/.task.backup
Project 'finance' is 95% complete (6 of 125 tasks remaining).

real	0m1.086s
user	0m1.069s
sys	0m0.017s
cbulle@debian-trixie:~$ time yes | task 10-20 mod wait:mon | ts
2026-06-27 06:34:35 This command will alter 11 tasks.
2026-06-27 06:34:35   - Wait will be changed from '9999-12-30' to '2026-06-29'.
2026-06-27 06:34:36 Modify task 10 'hidden'? (yes/no/all/quit) Modifying task 10 'hidden'.
2026-06-27 06:34:36 
2026-06-27 06:34:36   - Wait will be changed from '2027-08-30' to '2026-06-29'.
2026-06-27 06:34:37 Modify task 11 'hidden'? (yes/no/all/quit) Modifying task 11 'hidden'.
2026-06-27 06:34:37 
2026-06-27 06:34:37   - Wait will be changed from '9999-12-30' to '2026-06-29'.
2026-06-27 06:34:38 Modify task 12 'hidden'? (yes/no/all/quit) Modifying task 12 'hidden'.
2026-06-27 06:34:38 
2026-06-27 06:34:38   - Wait will be changed from '9999-12-30' to '2026-06-29'.
2026-06-27 06:34:39 Modify task 13 'hidden'? (yes/no/all/quit) Modifying task 13 'hidden'.
2026-06-27 06:34:39 
2026-06-27 06:34:39   - Wait will be changed from '9999-12-30' to '2026-06-29'.
2026-06-27 06:34:40 Modify task 14 'hidden'? (yes/no/all/quit) Modifying task 14 'hidden'.
2026-06-27 06:34:40 
2026-06-27 06:34:40   - Wait will be changed from '2026-05-03' to '2026-06-29'.
2026-06-27 06:34:41 Modify task 15 'hidden'? (yes/no/all/quit) Modifying task 15 'hidden'.
2026-06-27 06:34:41 
2026-06-27 06:34:41   - Wait will be changed from '2026-10-01' to '2026-06-29'.
2026-06-27 06:34:43 Modify task 16 'hidden'? (yes/no/all/quit) Modifying task 16 'hidden'.
2026-06-27 06:34:43 This is a recurring task.  Do you want to modify all pending recurrences of this same task? (yes/no) 
2026-06-27 06:34:43   - Wait will be changed from '2026-06-28' to '2026-06-29'.
2026-06-27 06:34:44 Modify task 17 'hidden'? (yes/no/all/quit) Modifying task 17 'hidden'.
2026-06-27 06:34:44 
2026-06-27 06:34:44   - Wait will be changed from '2026-12-01' to '2026-06-29'.
2026-06-27 06:34:45 Modify task 18 'hidden'? (yes/no/all/quit) Modifying task 18 'hidden'.
2026-06-27 06:34:45 
2026-06-27 06:34:45   - Wait will be changed from '2026-03-05' to '2026-06-29'.
2026-06-27 06:34:45 Modify task 19 'hidden'? (yes/no/all/quit) Modifying task 19 'hidden'.
2026-06-27 06:34:45 
2026-06-27 06:34:45   - Wait will be changed from '2026-06-28' to '2026-06-29'.
TASKDATA override: /tmp/.task.backup
Warning: You have specified that the 'wait' date is after the 'due' date.
Project '40c3' is 0% complete (1 task remaining).
Project 'finance' is 95% complete (6 of 125 tasks remaining).
Project 'freunde' is 98% complete (2 of 105 tasks remaining).
Project 'haushalt' is 99% complete (2 of 650 tasks remaining).
Project 'health' is 99% complete (1 of 403 tasks remaining).
Project 'job' is 98% complete (1 of 59 tasks remaining).
Project 'partner' is 97% complete (5 of 219 tasks remaining).
Project 'sec' is 96% complete (1 of 32 tasks remaining).
Project 'self' is 99% complete (1 of 162 tasks remaining).
2026-06-27 06:34:46 Modify task 20 'hidden'? (yes/no/all/quit) Modifying task 20 'hidden'.
2026-06-27 06:34:46 Modified 11 tasks.

real	0m10.578s
user	0m10.555s
sys	0m0.053s

cbulle@debian-trixie:~$ time task stats
TASKDATA override: /tmp/.task.backup

Category                   Data         
Pending                    2
Waiting                    32
Recurring                  2
Completed                  5555
Deleted                    1982
Total                      7573
Annotations                7953
Unique tags                11
Projects                   232
Blocked tasks              0
Blocking tasks             0
Undo transactions          850
Sync backlog transactions  84285
Tasks tagged               31.6%
Oldest task                2019-02-24
Newest task                2026-06-26
Task used for              7.3y
Task added every           8h
Task completed every       11h
Task deleted every         1d
Average time pending       12d
Average desc length        47 characters


real	0m1.230s
user	0m1.194s
sys	0m0.036s
```

</details>

I have provided the same data here: #4135 in case these are separate tickets
